stream-rpc/src/websocket.rs

```diff
@@ -0,0 +1,295 @@
+    //       tokio TcpListener can only be bound inside a tokio context.
+    let _guard = runtime.enter();
+    let routes = get_websocket_routes(config, diem_db);
+
+    let server = match &config.tls_cert_path {
+        None => Either::Left(warp::serve(routes).bind(config.address)),
+        Some(cert_path) => Either::Right(
+            warp::serve(routes)
+                .tls()
+                .cert_path(cert_path)
+                .key_path(config.tls_key_path.as_ref().unwrap())
+                .bind(config.address),
+        ),
+    };
+
+    runtime.handle().spawn(server);
+    runtime
+}
+
+#[derive(Clone)]
+pub struct ContextWrapperMapper {
+    pub context: ConnectionContext,
+}
+
+impl ContextWrapperMapper {
+    // Incoming messages
+    pub fn message_result_to_string(
+        &self,
+        result: Result<Message, warp::Error>,
+    ) -> Result<Option<String>, anyhow::Error> {
+        counters::MESSAGES_RECEIVED
+            .with_label_values(&[
+                self.context.transport.as_str(),
+                self.context.sdk_info.language.as_str(),
+                &self.context.sdk_info.version.to_string(),
+            ])
+            .inc();
+
+        match result {
+            Ok(msg) => {
+                // TODO: HANDLE PING + CLOSE MESSAGES
+                // TODO: DROP MESSAGES THAT ARE TOO BIG
+                match msg.to_str() {
+                    Ok(s) => Ok(Some(s.to_string())),
+                    // Just silently ignore messages that aren't text
+                    Err(_) => Ok(None),
+                }
+            }
+            Err(e) => Err(anyhow::Error::from(e)),
+        }
+    }
+    // Outgoing messages
+    fn string_result_to_message(
+        &self,
+        result: Result<String, anyhow::Error>,
+    ) -> Result<Message, warp::Error> {
+        counters::MESSAGES_SENT
+            .with_label_values(&[
+                self.context.transport.as_str(),
+                self.context.sdk_info.language.as_str(),
+                &self.context.sdk_info.version.to_string(),
+            ])
+            .inc();
+        match result {
+            Ok(item) => Ok(Message::text(item)),
+            // TODO: this aint right
+            Err(_e) => Ok(Message::close()),
+        }
+    }
+}
+
+pub async fn handle_websocket_stream(
+    user_agent: String,
+    socket: warp::ws::Ws,
+    cm: ConnectionManager,
+    config: Arc<WebsocketStreamConfig>,
+    headers: warp::http::HeaderMap,
+    remote_socket: Option<std::net::SocketAddr>,
+) -> Result<impl Reply, Rejection> {
+    let sdk_info = diem_json_rpc::util::sdk_info_from_user_agent(Some(&user_agent));
+    counters::HTTP_WEBSOCKET_REQUESTS
+        .with_label_values(&[sdk_info.language.as_str(), &sdk_info.version.to_string()])
+        .inc();
+
+    Ok(socket
+        .max_send_queue(config.send_queue_size as usize)
+        .max_message_size(config.content_length_limit as usize)
```

Comment:
Make sure the config types are what you're intending to use so that you don't need to use `as`